File: BQF/BQF-Src/BQFJR-Json5.3.1-DEV/RabbitMQClient/Common/JsonXmlObjectParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Xml;
using Newtonsoft.Json;

namespace SCM.RabbitMQClient.Common
{
    /// <summary>
    /// Json 转换
    /// </summary>
    public class JsonXmlObjectParser
    {
        /// <summary>
        /// Xml 转换Json
        /// </summary>
        /// <param name="xmlStr"></param>
        /// <returns></returns>
        public static string XmlToJson(string xmlStr)
        {
            var doc = new XmlDocument();
            doc.LoadXml(xmlStr);
            var json = JsonConvert.SerializeXmlNode(doc);
            return json;
        }
        /// <summary>
        /// Json 转换XMl
        /// </summary>
        /// <param name="jsonStr"></param>
        /// <returns></returns>
        public static string JsonToXml(string jsonStr)
        {
            var doc = (XmlDocument)JsonConvert.DeserializeXmlNode(jsonStr);
            return ConvertXmlToString(doc);
        }
        /// <summary>
        /// 以rootElementName为根的xml
        /// </summary>
        /// <param name="jsonStr"></param>
        /// <param name="rootElementName"></param>
        /// <returns></returns>
        public static string JsonToXml(string jsonStr, string rootElementName)
        {
            var doc = (XmlDocument)JsonConvert.DeserializeXmlNode(jsonStr, rootElementName);
            return ConvertXmlToString(doc);
        }
        /// <summary>
        /// 将XML 对象转换字符串
        /// </summary>
        /// <param name="doc"></param>
        /// <returns></returns>
        public static string ConvertXmlToString(XmlDocument doc)
        {
            var stream = new MemoryStream();
            var writer = new XmlTextWriter(stream, null);
            writer.Formatting = System.Xml.Formatting.Indented;

            doc.Save(writer);
            var sr = new StreamReader(stream, System.Text.Encoding.UTF8);
            stream.Position = 0;
            string xmlString = sr.ReadToEnd();
            sr.Close();
            stream.Close();
            return xmlString;
        }
        /// <summary>
        /// 将json 字符串转换成T对象
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="jsonStr"></param>
        /// <returns></returns>
        public static T DeserializeObject<T>(string jsonStr)
        {
            if (string.IsNullOrEmpty(jsonStr))
            {
                return default(T);
            }
            var jsonSerializerSettings = new JsonSerializerSettings();
            jsonSerializerSettings.NullValueHandling = NullValueHandling.Ignore;
            var jsonObject = JsonConvert.DeserializeObject<T>(jsonStr, jsonSerializerSettings);
            return jsonObject;
        }
        /// <summary>
        /// 某对象序列化
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public static string SerializeObject(object value)
        {
            if (value == null)
            {
                return "";
            }
            var jsonSerializerSettings = new JsonSerializerSettings();
            jsonSerializerSettings.NullValueHandling = NullValueHandling.Ignore;
            var objectJson = JsonConvert.SerializeObject(value, jsonSerializerSettings);
            return objectJson;
        }
        /// <summary>
        /// 某对象序列化为字节码
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public static byte[] SerializeToBytes(object value)
        {
            if (value == null)
            {
                return new byte[0];
            }
            var message = SerializeObject(value);
            var body = Encoding.UTF8.GetBytes(message);
            return body;
        }
        /// <summary>
        /// 某对象序列化为字节码
        /// </summary>
        /// <param name="value"></param>
        /// <returns></returns>
        public static byte[] SerializeToBytes(object value, Encoding encoder)
        {
            if (value == null)
            {
                return new byte[0];
            }
            var message = SerializeObject(value);
            var body = encoder.GetBytes(message);
            return body;
        } 
        /// <summary>
        /// 将某字节码反序列化对象T
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="value"></param>
        /// <returns></returns>
        public static T DeserializeBytes<T>(byte[] value)
        {
            if (value == null)
            {
                return default(T);
            }
            var result = Encoding.UTF8.GetString(value);
            var tObject = DeserializeObject<T>(result);
            return tObject;
        }
        /// <summary>
        /// 将某字节码反序列化对象T
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="value"></param>
        /// <returns></returns>
        public static T DeserializeBytes<T>(byte[] value,Encoding encoder)
        {
            if (value == null)
            {
                return default(T);
            }
            var result = encoder.GetString(value);
            var tObject = DeserializeObject<T>(result);
            return tObject;
        }
    }
}

File: BQF/BQF-Src/BQFJR-Json5.3.1-Release/Instinct.Windows.Service.ActionService/JsonXmlObjectParser.vb
Imports Newtonsoft.Json
Imports System.IO
Imports System.Xml
Imports System.Text

Public Class JsonXmlObjectParser
    ''' <summary>
    ''' Convert XML to JSON
    ''' </summary>
    ''' <param name="StrXml"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Shared Function XmlToJson(ByVal StrXml As String) As String
        Dim doc As New Xml.XmlDocument
        doc.LoadXml(StrXml)
        Dim json As String = JsonConvert.SerializeXmlNode(doc)
        Return json
    End Function
    ''' <summary>
    ''' Convert JSON to XML
    ''' </summary>
    ''' <param name="StrJson"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Shared Function JsonToXml(ByVal StrJson) As String
        Dim doc As Xml.XmlDocument
        doc = JsonConvert.DeserializeXmlNode(StrJson)
        Return ConvertXmlToString(doc)
    End Function
    ''' <summary>
    ''' Convert XMl Object TO String
    ''' </summary>
    ''' <param name="doc"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Shared Function ConvertXmlToString(ByVal doc As XmlDocument) As String

        Dim stream As New MemoryStream()
        Dim writer As New XmlTextWriter(stream, Nothing)
        writer.Formatting = System.Xml.Formatting.Indented

        doc.Save(writer)
        Dim sr As New StreamReader(stream, System.Text.Encoding.UTF8)
        stream.Position = 0
        Dim xmlString As String = sr.ReadToEnd()
        sr.Close()
        stream.Close()
        Return xmlString

    End Function

End Class
